backend/src/apiserver/resource/resource_manager.go

```
@@ -153,6 +156,88 @@ func NewResourceManager(clientManager ClientManagerInterface, options *ResourceM
 	}
 }
 
+// extractPipelineRunParallelism extracts pipeline_run_parallelism from the template's platform spec.
+// It navigates: template -> platformSpec -> platforms["kubernetes"] -> pipelineConfig -> pipelineRunParallelism
+func extractPipelineRunParallelism(tmpl template.Template) (int64, bool) {
+	if tmpl == nil || !tmpl.IsV2() {
+		return 0, false
+	}
+	// Type assertion: cast to V2Spec to access PlatformSpec()
+	v2Spec, ok := tmpl.(*template.V2Spec)
+	if !ok || v2Spec == nil {
+		return 0, false
+	}
+	// Get the platform spec (already public method)
+	platformSpec := v2Spec.PlatformSpec()
+	if platformSpec == nil {
+		return 0, false
+	}
+	// Get kubernetes platform config
+	kubernetesSpec, ok := platformSpec.Platforms["kubernetes"]
+	if !ok || kubernetesSpec == nil {
+		return 0, false
+	}
+	// Get pipeline config
+	pipelineConfig := kubernetesSpec.GetPipelineConfig()
+	if pipelineConfig == nil || pipelineConfig.PipelineRunParallelism == nil {
+		return 0, false
+	}
+	// Get the parallelism value
+	value := pipelineConfig.GetPipelineRunParallelism()
+	if value <= 0 {
+		return 0, false
```

We should return an error here.